superset-frontend/src/middleware/asyncEvent.ts

```diff
@@ -0,0 +1,194 @@
+    } catch (err) {
+      console.warn('failed to fetch last event Id from localStorage');
+    }
+
+    const fetchEvents = makeApi<
+      { last_id?: string | null },
+      { result: AsyncEvent[] }
+    >({
+      method: 'GET',
+      endpoint: POLLING_URL,
+    });
+
+    const fetchCachedData = async (
+      asyncEvent: AsyncEvent,
+      componentId: number,
+    ): Promise<CachedDataResponse> => {
+      let status = 'success';
+      let data;
+      try {
+        const { json } = await SupersetClient.get({
+          endpoint: asyncEvent.result_url,
+        });
+        data = 'result' in json ? json.result[0] : json;
+      } catch (response) {
+        status = 'error';
+        data = await getClientErrorObject(response);
+      }
+
+      return { componentId, status, data };
+    };
+
+    const setLastId = (asyncEvent: AsyncEvent) => {
+      lastReceivedEventId = asyncEvent.id;
+      try {
+        localStorage.setItem(LOCALSTORAGE_KEY, lastReceivedEventId as string);
+      } catch (err) {
+        console.warn('Error saving event ID to localStorage', err);
+      }
+    };
+
+    const processEvents = async () => {
+      const state = store.getState();
+      const queuedComponents = getPendingComponents(state);
+      const eventArgs = lastReceivedEventId
+        ? { last_id: lastReceivedEventId }
+        : {};
+      const events: AsyncEvent[] = [];
+      if (queuedComponents && queuedComponents.length) {
+        try {
+          const { result: events } = await fetchEvents(eventArgs);
+          if (events && events.length) {
+            const componentsByJobId = queuedComponents.reduce((acc, item) => {
+              acc[item.asyncJobId] = item;
+              return acc;
+            }, {});
+            const fetchDataEvents: Promise<CachedDataResponse>[] = [];
+            events.forEach((asyncEvent: AsyncEvent) => {
+              const component = componentsByJobId[asyncEvent.job_id];
+              if (!component) {
+                console.warn(
+                  'component not found for job_id',
+                  asyncEvent.job_id,
+                );
+                return setLastId(asyncEvent);
+              }
+              const componentId = component.id;
+              switch (asyncEvent.status) {
+                case JOB_STATUS.DONE:
+                  fetchDataEvents.push(
+                    fetchCachedData(asyncEvent, componentId),
+                  );
+                  break;
+                case JOB_STATUS.ERROR:
+                  store.dispatch(
+                    errorAction(componentId, parseErrorJson(asyncEvent)),
+                  );
+                  break;
+                default:
+                  console.warn('received event with status', asyncEvent.status);
+              }
+
+              return setLastId(asyncEvent);
+            });
+
+            const fetchResults = await Promise.all(fetchDataEvents);
```

Comment:
Wondering if it's most performant to do an `all` here... This will cause all results to be loaded at once into memory as data is fetched (not that it doesn't already after landing in the DOM), but from a GC POV, it might be better to just iterate over your collected promises and resolve each, as you go

Reply:
The performance gain in `Promise.all`'s request parallelization is a major factor here. If we run into memory issues down the road, we can revisit this. Ideally, the aggregated data being returned to charts is not huge, but I know that's not always the case.